### PR TITLE
Solution creation flow

### PR DESCRIPTION
**Pull request by praveenKDass**

### CODE REVIEW

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed 8 files added comments based on that

File: controllers/v1/programsController.js [hunk 1]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
   * @apiUse errorBody
   */
 
-  /**
+   /**
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:57]: Changed the URl as /samiksha/v1/programs/ in all the places @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 2]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
-   */
+   * @param {Object} req - Requested data.
+   * @param {Array} req.query.page - Page number.
+   * @param {Array} req.query.limit - Page Limit.
+   * @param {Array} req.query.search - Search text.
+   * @returns {JSON} List programs data.
+  */
 
+
+  // async list(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:69]: Removed the commented code

File: controllers/v1/programsController.js [hunk 3]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+
+  //       let response = { message: messageConstants.apiResponses.PROGRAM_LIST, result: programDocument };
+
+  //       return resolve(response);
+  //     } catch (error) {
+  //       return reject({ message: error });
+  //     }
+  //   });
+  // }
   async list(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:109]: For list not required validator so didn't add for it but for other program APIs added the validator @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 4]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
-              description: 1,
-              'assessments._id': 1,
-              'assessments.externalId': 1,
-              'assessments.name': 1,
-              'assessments.description': 1,
-            },
-          },
-        ]);
+        let listOfPrograms = await programsHelper.list(
+          req.pageNo,
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:113]: So middleware will convert req.query.page to req.pageNo and same for limit so keep it like that and also while testing the API pagination worked correctly @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 5]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+      } catch (error) {
+        return reject({
+          status: error.status || httpStatusCode.internal_server_error.status,
+          message: error.message || httpStatusCode.internal_server_error.message,
+          errorObject: error
+        });
+      }
+    });
+  }
+
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:132]: Added api doc @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 6]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+ 
+   /**
+   * Create program.
+   * @method
+   * @name create
+   * @param {Object} req - requested data.
+   * @returns {JSON} - created program document.
+   */
+
+   async create(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:141]: Added validator for create @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 7]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+    return new Promise(async (resolve, reject) => {
+      try {
+
+        req.body.userId = req.userDetails.userId;
+        let programCreationData = await programsHelper.create(
+          req.body,
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:147]: added the comment @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 8]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+          status: error.status || httpStatusCode.internal_server_error.status,
+          message: error.message || httpStatusCode.internal_server_error.message,
+          errorObject: error
+        })
+      }
+    })
+  }
+  
+
+   /**
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:169]: Added API document for update @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 9]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+   * Update program.
+   * @method
+   * @name update
+   * @param {Object} req - requested data.
+   * @param {Object} 
+   * @returns {JSON} - 
+   */
+
+   async update(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:178]: added validator @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 10]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+    return new Promise(async (resolve, reject) => {
+      try {
+
+        let programUpdationData = await programsHelper.update(
+          req.params._id,
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:183]: Added comment for boolean value which passing in arg @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 11]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+        reject({
+          status: error.status || httpStatusCode.internal_server_error.status,
+          message: error.message || httpStatusCode.internal_server_error.message,
+          errorObject: error
+        })
+      }
+    })
+  }
+
+
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:204]: added api document @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 12]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+   * Add roles in program scope
+   * @method
+   * @name addRolesInScope
+   * @param {Object} req - requested data.
+   * @param {String} req.params._id - program id.
+   * @param {Array} req.body.roles - Roles to be added.
+   * @returns {Array} Program scope roles.
+   */
+
+    async addRolesInScope(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:215]: added validator for  addRolesInScope Api @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 13]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+            status: error.status || httpStatusCode.internal_server_error.status,
+            message: error.message || httpStatusCode.internal_server_error.message,
+            errorObject: error
           });
         }
+      });
+    }
 
-        let response = { message: messageConstants.apiResponses.PROGRAM_LIST, result: programDocument };
+        /**
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:236]: added @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 14]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+   * Add entities in program scope
+   * @method
+   * @name addEntitiesInScope
+   * @param {Object} req - requested data.
+   * @param {String} req.params._id - program id.
+   * @param {Array} req.body.entities - Entities to be added.
+   * @returns {Array} Program scope roles.
+   */
+
+  async addEntitiesInScope(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:246]: added the validator @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 15]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+        return reject({
+          status: error.status || httpStatusCode.internal_server_error.status,
+          message: error.message || httpStatusCode.internal_server_error.message,
+          errorObject: error
+        });
       }
     });
   }
 
+     /**
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:268]: Added api document @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 16]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+   * Remove roles in program scope
+   * @method
+   * @name removeRolesInScope
+   * @param {Object} req - requested data.
+   * @param {String} req.params._id - program id.
+   * @param {Array} req.body.roles - Roles to be added.
+   * @returns {Array} Program scope roles.
+   */
+
+     async removeRolesInScope(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:278]: added validator @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 17]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+            status: error.status || httpStatusCode.internal_server_error.status,
+            message: error.message || httpStatusCode.internal_server_error.message,
+            errorObject: error
+          });
+        }
+      });
+    }
+
+     /**
+    * @api {post} /assessment/api/v1/programs/removeEntitiesInScope/:programId Add roles in programs
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:300]: changed to  /samiksha/v1/programs/removeEntitiesInScope @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 18]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+   * Remove entities in program scope
+   * @method
+   * @name removeEntitiesInScope
+   * @param {Object} req - requested data.
+   * @param {String} req.params._id - program id.
+   * @param {Array} req.body.entities - Entities to be added.
+   * @returns {Array} Program scope roles.
+   */
+
+  async removeEntitiesInScope(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:329]: its required and added in program module validator file @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 19]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+          message: error.message || httpStatusCode.internal_server_error.message,
+          errorObject: error
+        });
+      }
+    });
+  }
+
+ 
+  /**
+    * @api {get} /assessment/api/v1/programs/details/:programId Program Details
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:352]: Changed url to  /samiksha/v1/programs/details @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 20]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+     /**
+   * Details of the program
+   * @method
+   * @name details
+   * @param {Object} req - requested data.
+   * @param {String} req.params._id - program id.
+   * @returns {Array} Program scope roles.
+   */
+
+     async details(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:379]: added validator @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 21]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+    /**
+   * join program
+   * @method
+   * @name join
+   * @param {Object} req - requested data.
+   * @param {String} req.params._id - program id.
+   * @returns {Object} Program join status.
+  */
+
+   async join(req) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:410]: Added validator for join api @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 22]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+    return new Promise(async (resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:411]: Only for this function and details removed that and used only async await method @VISHNUDAS-tunerlabs ..its working as expected if we want keep that way in future we will create a new ticket for that and will make the change..let me know your idea or if need any changes @VISHNUDAS-tunerlabs

File: controllers/v1/programsController.js [hunk 23]
```diff
@@ -54,54 +54,387 @@ module.exports = class Programs extends Abstract {
...
+      } catch (error) {
+        return reject({
+          status: error.status || httpStatusCode.internal_server_error.status,
+          message: error.message || httpStatusCode.internal_server_error.message,
+          errorObject: error
+        });
+      }
+    });
+  }
   /**
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply controllers/v1/programsController.js:441]: Changed URLs @VISHNUDAS-tunerlabs

File: generics/messageConstants/endpoints.js [hunk 24]
```diff
@@ -17,4 +17,9 @@ module.exports = {
   SOLUTION_DETAILS_BASED_ON_ROLE_LOCATION: 'api/v1/solutions/detailsBasedOnRoleAndLocation',
   GET_PROJECT_TEMPLATE_LISTS: 'api/v1/project/templates/listByIds',
   DOWNLOADABLE_FILE_URL: 'api/v1/cloud-services/files/getDownloadableUrl',
+  GET_LOCATION_DATA: '/v1/location/search',
+  GET_SCHOOL_DATA: '/v1/org/search',
+  USER_CONSENT_API: '/v1/user/consent/update',
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply generics/messageConstants/endpoints.js:24]: No we were only using  GET_LOCATION_DATA,and  USER_READ_V5 and  USER_CONSENT_API on this i have some doubts will connect over the call for this @VISHNUDAS-tunerlabs

File: generics/services/users.js [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+ * @param {String} pageSize -  requested page size.
+ * @param {String} pageNo -  requested page number.
+ * @param {String} searchKey -  search string.
+ * @param {Boolean} formatResult - format result
+ * @param {Boolean} returnObject - return object or array.
+ * @param {Boolean} resultForSearchEntities - format result for searchEntities api call.
+ * @returns {Promise} returns a promise.
+ */
+
+const locationSearch = function (
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply generics/services/users.js:80]: Yes moved to entity-management service @VISHNUDAS-tunerlabs

File: generics/services/users.js [hunk 26]
```diff
@@ -0,0 +1,386 @@
+//dependencies
+const request = require('request');
+const userServiceUrl = process.env.USER_SERVICE_URL;
+const entityServiceUrl = process.env.ENTITY_SERVICE_URL
+
+
+
+const profile = function ( token,userId = "" ) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply generics/services/users.js:8]: Yes @VISHNUDAS-tunerlabs .its working i can able to fetch the user info

File: generics/services/users.js [hunk 27]
```diff
@@ -0,0 +1,386 @@
...
+ * @name orgSchoolSearch
+ *  @param {object} filterData -  contain filter object.
+ * @param {String} pageSize -  requested page size.
+ * @param {String} pageNo -  requested page number.
+ * @param {String} searchKey -  search string.
+ * @param {String} searchKey - search key for fuzzy search.
+ * @param {String} fields -  required field filter.
+ * @returns {Promise} returns a promise.
+ */
+const orgSchoolSearch = function (filterData, pageSize = '', pageNo = '', searchKey = '', fields = []) {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply generics/services/users.js:218]: Yes we were not using this API call  But i need a small doubt over that. will connect for this and endpoint comment @VISHNUDAS-tunerlabs

File: module/programs/helper.js [hunk 28]
```diff
@@ -7,50 +7,187 @@
...
+          });
+        }
+        if (pageNo !== '' && pageSize !== '') {
+          pagination = {
+            skip: pageSize * (pageNo - 1),
+            limit: pageSize,
+          };
+        }
+
+        let programData = await database.models.programs.find(queryObject, projection, pagination).lean();
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply module/programs/helper.js:58]: Added a databaseQuery @VISHNUDAS-tunerlabs

File: module/programs/helper.js [hunk 29]
```diff
@@ -7,50 +7,187 @@
...
+   * @param {Array} [filterQuery = "all"] - solution ids.
+   * @param {Array} [fieldsArray = "all"] - projected fields.
+   * @param {Array} [skipFields = "none"] - field not to include.
+   * @param {Number} pageNo - page no.
+   * @param {Number} pageSize - page size.
+   * @returns {Array} List of programs.
+   */
+
+  static programDocuments(filterQuery = 'all', fieldsArray = 'all', skipFields = 'none', pageNo = '', pageSize = '') {
+    return new Promise(async (resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply module/programs/helper.js:34]: moved it to databaseQuery file @VISHNUDAS-tunerlabs

File: module/programs/helper.js [hunk 30]
```diff
@@ -141,15 +278,860 @@ module.exports = class ProgramsHelper {
...
+          }
+          if (data.hasOwnProperty('startDate')) {
+            data.startDate = gen.utils.getStartDate(data.startDate, timeZoneDifference);
+          }
+        }
+
+        _.assign(programData, {
+          ...data,
+        });
+        programData = _.omit(programData, ['scope', 'userId']);
```
[the remark this answers is not in the document]
Comment (praveenKDass) [reply module/programs/helper.js:293]: intergrated as mentioned @VISHNUDAS-tunerlabs

**Pull request closed without merging**